### PR TITLE
dependency-extraction-webpack-plugin: Calculate vendor hash from file output rather than Webpack internal state

### LINKED ISSUE #34660: `@wordpress/dependency-extraction-webpack-plugin` should use content hash
### Description

`@wordpress/dependency-extraction-webpack-plugin` outputs the equivalent of `[fullhash]`/`[chunkhash]` as the "version" rather than `[contenthash]`. Note that Webpack itself [now recommends using `[contenthash]`](https://github.com/webpack/webpack.js.org/issues/2096).

The use of `[fullhash]`/`[chunkhash]` can result in extraneous hash changes. For example, if the JavaScript is identical but some CSS changed, it will still change the hash for the JavaScript files.

In certain configurations, this can even result in a hash change even if the code is unchanged, if the code is built in a different directory. See the reproduction below. We've been running into this frequently with the wpcom TeamCity builds of Jetpack.

You might fix this by changing
https://github.com/WordPress/gutenberg/blob/eb4c8e38f9bac57494b782f7699a5538a35dd6b8/packages/dependency-extraction-webpack-plugin/lib/index.js#L186
to
```js
version: runtimeChunk.contentHash && runtimeChunk.contentHash.javascript || runtimeChunk.hash,
```

### Step-by-step reproduction instructions

1. Download and extract [test.zip](https://github.com/WordPress/gutenberg/files/7129860/test.zip)
2. Run `npm install && npx webpack`. It also works with `pnpm`; I haven't tested `yarn`.
3. Extract test.zip again to a different path.
4. Run `npm install && npx webpack` there too.
5. Compare the contents of the `dist` directories. See that, although everything else is identical, the "version" hash in the asset.php file is different.

What's going on here is that `@babel/plugin-transform-runtime` with `absoluteRuntime: true` winds up injecting a variable name derived from the absolute path to `@babel/runtime`. Even though this variable name gets minified away, it still affects the `[fullhash]`/`[chunkhash]`.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

N/A

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by anomiex**
## Description

Webpack 5 recommends using the content hash over the old-style chunk
hash for cache invalidation, as the chunk hash can change even if the
output does not.

dependency-extraction-webpack-plugin will now include the content hash
for each content type in the asset file. The `version` field is still
present, now as a combined hash for all the content types reported, to
maintain backwards compatibility.

Fixes #34660

## How has this been tested?

Tested locally, both in a trivial repo and in https://github.com/Automattic/jetpack/tree/master/projects/plugins/jetpack.

## Screenshots 
N/A

## Types of changes

New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by spacedmonkey** [comment]
Wouldn't this be a breaking change?

**Comment by anomiex** [comment]
What do you think it would break?

**Comment by gziolo** [comment]
@anomiex, thank you for opening this PR.

For backward compatibility, we should optimize that the usage of `version` in the first place. It's used in WordPress core here:

https://github.com/WordPress/wordpress-develop/blob/d802fecf979d90b568e3c84d17d521df200cab38/src/wp-includes/blocks.php#L109-L114

The current behavior is also documented in the following section in the Block Editor Handbook:

https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#wpdefinedasset ([source](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md#wpdefinedasset))

I would expect that many plugins depend on that, too. 

What is the rationale to have a different value for `version` and `contentHash.javaScript`? Why developers would want to use the new version? I like that you could now get the way to automate versioning for other resources. How could it be used in practice by plugins?

**Comment by anomiex** [comment]
> What is the rationale to have a different value for `version` and `contentHash.javaScript`?

The `version` encompasses all the assets for backwards compatibility. `contentHash.javaScript` covers only the JavaScript.

We could go to some extra work if we really wanted to detect the case where the only asset is JavaScript and make the two hashes be the same instead of calculating a combined hash. It doesn't seem worth the effort to me, though.

> Why developers would want to use the new version? I like that you could now get the way to automate versioning for other resources. How could it be used in practice by plugins?

As noted in the patched readme, the JavaScript hash could be passed to `wp_enqueue_script()` while the CSS hash could be passed to `wp_enqueue_style()`, so if a change is made that only affects the CSS then the JS cache wouldn't have to be busted too.

On the other hand, they could just keep using `version` for simplicity. The main point of this for me is for the hashes to stop changing when none of the actual content changed, as detailed in #34660.

**Comment by gziolo** [comment]
From my perspective, it should be perfectly fine to avoid introducing `contentHash.javaScript` and change the behavior of `version`. The changes in the documentation favor `contentHash.javaScript` over `version`, so there wouldn't be too much value in keeping `version` other than for backward compatibility. It would be also confusing to have an old and a new way to invalidate the cache for JavaScript assets. How plugin authors would know which one to use?

**Comment by anomiex** [comment]
Are you requesting changes to the PR, or just musing?

**Comment by gziolo** [comment]
I think that this PR needs some iterations to avoid a need for updating PHP code to use the updated hash for JS files.

**Comment by anomiex** [comment]
There's no _need_ as it is now, it's fully backwards compatible. But if you want a revision to make it so there's no code update _possible_, that's fine with me.

On the other hand, as I implemented that and re-tested I found out my code wasn't working right. Webpack calculates the chunk contentHash fields when it first processes the chunk, but then when further updates happen (e.g. minification) it only updates the hash where it appears in the assets. It turns out that it worked with the code snippet from #34660 that only looked at the JavaScript asset, but fails when we combine all the assets' hashes.

**Comment by anomiex** [comment]
@gziolo It has been five months since I made the change you requested. Care to re-review?

**Comment by gziolo** [comment]
> @gziolo It has been five months since I made the change you requested. Care to re-review?

I'll ask for help with the review in the parent issue. I don't think I fully understand the issue here.

**Comment by gziolo** [comment]
> @gziolo It has been five months since I made the change you requested. Care to re-review?

I'll ask for help with the review in the parent issue. I don't think I fully understand the issue here.

**Comment by kraftbj** [comment]
Confirming that this PR resolves the issue that we're facing.

The full narrative of why it matters to us (Jetpack):

- we use the Gutenberg-developed dependency-extraction-webpack-plugin package to help us manage WordPress script dependencies within Jetpack-developed blocks.

- As expected, it creates a PHP file at build time to include the WordPress script as a dependency in WP friendly way. For example, https://github.com/Automattic/jetpack-production/blob/master/_inc/blocks/business-hours/view.asset.php

- The problem is the version field there. The “bug” is that the version field can change even if the underlying code we’re shipping remains the same. TBH, this doesn’t matter to most people. They’ll just accept it and move on.

- For us, though, we’re also trying to sync these blocks to the WP.com codebase, which currently we’re taking the same source code and is built to meet [WP.com](http://wp.com/)’s needs. The impact is the same files building the same results can end up with different hashes.

- In an example, the only actual change originally was changing the stable tab in readme.txt. Nothing else changed. However, because of the way that version is constructed right now, it uses the full path (e.g. /home/users/kraft/code/something/something/ as part of the math to figure out the version hash, even though it isn’t needed). The original issue provides an easy reproduction example: https://github.com/WordPress/gutenberg/issues/34660

- So, since WP.com is building the files in different directories (depending on which TeamCity host is executing the build), we get in situations where a random diff will be forced to include a pointless version bump.

- In and of itself, that probably isn’t a huge deal, but we run into a conflict where we have two diffs go in at the same time. Neither one of them requires anything changing to the end result of the build, but both trigger this. Now we run into merge conflicts after one lands, so we need to force [WP.com](http://wp.com/) to rebuild to avoid the merge conflict.

- The solution here is to tweak the method used to use what Webpack is now suggesting to use anyhow, `contenthash` which is a version number determined based on the output of the build itself… not something like the location of the files, etc. This resolves the issue we're seeing.

**Comment by gziolo** [comment]
> In an example, the only actual change originally was changing the stable tab in readme.txt. Nothing else changed. However, because of the way that version is constructed right now, it uses the full path (e.g. /home/users/kraft/code/something/something/ as part of the math to figure out the version hash, even though it isn’t needed). The original issue provides an easy reproduction example: https://github.com/WordPress/gutenberg/issues/34660

We had exactly the same issue in WordPress core a year ago. @peterwilsoncc patched it by using `optimization.moduleIds` set to `hashed` (webpack v4, `deterministic` in v5`). See https://core.trac.wordpress.org/ticket/53192. Anyway, I will look closer into the attached project and try it myself and see whether changing the webpack config takes any effect.

**Comment by gziolo** [comment]
> > In an example, the only actual change originally was changing the stable tab in readme.txt. Nothing else changed. However, because of the way that version is constructed right now, it uses the full path (e.g. /home/users/kraft/code/something/something/ as part of the math to figure out the version hash, even though it isn’t needed). The original issue provides an easy reproduction example: #34660
> 
> We had exactly the same issue in WordPress core a year ago. @peterwilsoncc patched it by using `optimization.moduleIds` set to `hashed` (webpack v4, `deterministic` in v5`). See https://core.trac.wordpress.org/ticket/53192. Anyway, I will look closer into the attached project and try it myself and see whether changing the webpack config takes any effect.

Ok, it looks like `deterministic` is the default setting in webpack 5, so it won’t help here.

It looks like the hash (`version`) generated for the asset depends on the absolute path of the entry point. I was able to confirm that before this patch moving the reproduction project to a different folder changes the hash generated. When it gets moved back then the hash gets restored.

File: packages/dependency-extraction-webpack-plugin/lib/index.js [hunk 1]
```diff
@@ -198,14 +198,25 @@ class DependencyExtractionWebpackPlugin {
 				}
 			}
 
+			// Go through the assets and hash the sources. We can't just use
+			// `entrypointChunk.contentHash` because that's not updated when
+			// assets are minified. Sigh.
+			// @todo Use `asset.info.contenthash` if we can make sure it's reliably set.
```
Comment (gziolo) [inline packages/dependency-extraction-webpack-plugin/lib/index.js:204]: Does it mean that it could get fixed on the webpack side? Is there an issue open that describes the same use case and could be included as a reference so this todo item can be revisited later?

File: packages/dependency-extraction-webpack-plugin/lib/index.js [hunk 2]
```diff
@@ -198,14 +198,25 @@ class DependencyExtractionWebpackPlugin {
...
+			const hash = createHash( 'sha512' );
+			for ( const filename of entrypoint.getFiles().sort() ) {
+				const asset = compilation.getAsset( filename );
+				hash.update( `${ filename }: ` );
```
Comment (gziolo) [inline packages/dependency-extraction-webpack-plugin/lib/index.js:208]: Does it mean that the the hash generated in the asset file depends now only on the list of sorted file names imported in the entry?

File: packages/dependency-extraction-webpack-plugin/lib/index.js (hunk 1 above)
Comment (anomiex) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:204]: No, that's not what it means.

What this comment is referring to is that _sometimes_ Webpack sets `.info.contenthash` on the `asset` object, but it only does so if it decides that it needs it (if I recall correctly, it depends on whether the filename template uses `[contenthash]`). If we could ensure that Webpack sets that every time, we could just use it instead of hashing the file contents ourself.

File: packages/dependency-extraction-webpack-plugin/lib/index.js (hunk 2 above)
Comment (anomiex) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:208]: No. See the very next line.

Comment (gziolo) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:208]: What’s the next line doing then? I don’t know internal webpack API.

Comment (anomiex) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:208]: Calling `hash.update()` with the contents of the asset.

Comment (anomiex) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:208]: Now that I think about it I can see where your confusion may have come from, it's kind of backwards that `asset.source.updateHash( hash )` updates the passed `hash` rather than the `asset.source`, and https://www.npmjs.com/package/webpack-sources#updateHash doesn't seem to come up in the Google searches I tried (I only found it just now when grepping the code for docs).

File: packages/dependency-extraction-webpack-plugin/lib/index.js [hunk 3]
```diff
@@ -198,14 +198,25 @@ class DependencyExtractionWebpackPlugin {
...
+				asset.source.updateHash( hash );
```
Comment (gziolo) [inline packages/dependency-extraction-webpack-plugin/lib/index.js:209]:
```suggestion
				// Updates the provided Hash object with the content of the represented source code.
				asset.source.updateHash( hash );
```

File: packages/dependency-extraction-webpack-plugin/lib/index.js (hunk 2 above)
Comment (gziolo) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:208]: Thank you. Now it all makes sense. I tested it and it works perfectly fine.

I tested with imported CSS files and changes to the content impact the hash generated for the JS file, but it's similar to the current behavior when using `entrypointChunk.hash`. I guess that's fine.

File: packages/dependency-extraction-webpack-plugin/lib/index.js [hunk 4]
```diff
@@ -198,14 +198,25 @@ class DependencyExtractionWebpackPlugin {
 				}
 			}
 
+			// Go through the assets and hash the sources. We can't just use
+			// `entrypointChunk.contentHash` because that's not updated when
+			// assets are minified. Sigh.
+			// @todo Use `asset.info.contenthash` if we can make sure it's reliably set.
+			const hash = createHash( 'sha512' );
+			for ( const filename of entrypoint.getFiles().sort() ) {
+				const asset = compilation.getAsset( filename );
```
Comment (gziolo) [inline packages/dependency-extraction-webpack-plugin/lib/index.js:207]: It looks like we could simplify the code here to:

```js
hash.update( filename + ':' + compilation.getAsset( filename ).source.source() );
```

Would there be any drawbacks?

File: packages/dependency-extraction-webpack-plugin/lib/index.js [hunk 5]
```diff
@@ -198,14 +198,25 @@ class DependencyExtractionWebpackPlugin {
...
+				hash.update( `${ filename }: ` );
+				asset.source.updateHash( hash );
+			}
+
 			const entrypointChunk = isWebpack4
```
Comment (gziolo) [inline packages/dependency-extraction-webpack-plugin/lib/index.js:212]: ~We can now move `entrypointChunk` constant further in the code and closer to the usage.~ Disregard this one.

Comment (gziolo) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:212]: Actually, we probably could get files from the `entrypointChunk.files` rather than from `entrypoint.getFiles()` to align with the handling in other places:

```js
const entrypointChunk = isWebpack4
	? entrypoint.chunks.find( ( c ) => c.name === entrypointName )
	: entrypoint.getEntrypointChunk();

const entrypointChunkHash = createHash( 'sha512' );
for ( const filename of Array.from( entrypointChunk.files ).sort() ) {
	entrypointChunkHash.update( filename + ':' + compilation.getAsset( filename ).source.source() );
}
```

File: packages/dependency-extraction-webpack-plugin/lib/index.js (hunk 1 above)
Comment (gziolo) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:204]: Got it. I think the algorithm you proposed is a nice improvement so let's rephrase the comment so it presents the benefits, instead of the limitations of the webpack. I can fully confirm that it's going to solve the issue that Jetpack struggles with as explained by @kraftbj with hashes changing depending on the absolute path for files. We can also link this PR so people can learn the full context.

**Review by gziolo — COMMENTED** [review]
I have tested this PR extensively using the project scaffolded with `npx @wordpress/create-block -t @wordpress/create-block-tutorial-template` and all the changes were applied directly to the `node_modules` folder. I can confirm it works as expected and fixes the issue described by @kraftbj. I left some comments to discuss the current implementation, but overall we can land this change.

We should also update the title for this PR because it doesn't reflect the current implementation.

We maintain the CHANGELOG files for the package, and it would be great to list this change as well because it will change `version` in all generated assets files with this change included:
https://github.com/WordPress/gutenberg/blob/trunk/packages/dependency-extraction-webpack-plugin/CHANGELOG.md

File: packages/dependency-extraction-webpack-plugin/lib/index.js (hunk 5 above)
Comment (gziolo) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:212]: In `@wordpress/scripts` there is this logic:

https://github.com/WordPress/gutenberg/blob/67fcf07c2da488f1e5ff9bc0dda648388911b35c/packages/scripts/config/webpack.config.js#L107-L123

With the current implementation it processes 3 files:

```
[ './style-index.css', 'index.css', 'index.js' ]
```

`Array.from( entrypointChunk.files )` would process only 2 files because `./style-index.css` goes to its own chunk:

```
[ 'index.css', 'index.js' ]
```

File: packages/dependency-extraction-webpack-plugin/lib/index.js (hunk 4 above)
Comment (anomiex) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:207]: Only in that `asset.source.updateHash()` may be more efficient, processing the internal representation of the source directly instead of having to stringify it for concatenation first.

File: packages/dependency-extraction-webpack-plugin/lib/index.js (hunk 1 above)
Comment (anomiex) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:204]: I have no objection to changing the comment, but I don't know what to rephrase it to. Care to make a suggestion?

File: packages/dependency-extraction-webpack-plugin/lib/index.js (hunk 4 above)
Comment (gziolo) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:207]: Right, that could be the case.

File: packages/dependency-extraction-webpack-plugin/lib/index.js (hunk 5 above)
Comment (anomiex) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:212]: Are you still suggesting to change the implementation here as in your second comment, or is the third comment convincing yourself not to?

It seems to me that your last comment is the argument not to make the change you suggested in your second comment. The extra file is part of the asset and so should be covered by the hash.

Comment (gziolo) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:212]: My last comment provides the reasoning why the change included in https://github.com/WordPress/gutenberg/pull/34969#discussion_r865656919 would be a good improvement. We don't care about `./style-index.css` that goes into its chunk so it shouldn't matter when calculating the hash for the JS entry point.

Comment (anomiex) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:212]: Would you care if I point out that someone using `.optimization.runtimeChunk` would find that the `runtime.js` or `runtime~index.js` isn't included in the version hash either?

If you really want this change I'm not going to fight it since our use case has neither of these sorts of extra files. But I do think if we're going to have one hash for the asset, it should cover the whole asset rather than excluding parts of it.

Comment (anomiex) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:212]: There's also the possibility that someone is using splitChunks like you do there but on JS code chunks, or with Webpack's automatic vendor splitting.

File: packages/dependency-extraction-webpack-plugin/lib/index.js (hunk 1 above)
Comment (jsnajdr) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:204]: If I understand it correctly, ideally we would just change (line 208):
```
version: entrypointChunk.hash
```
to:
```
version: entrypointChunk.contenthash
```
and be done? But we can't do that because webpack doesn't always set it? And/or the value doesn't have the properties that we need?

I noticed that webpack has an [`optimization.realContentHash`](https://webpack.js.org/configuration/optimization/#optimizationrealcontenthash) config option. Which means that there are multiple valid ways how to compute `contenthash`, and that the default ways is in some sense "not real."

Comment (anomiex) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:204]: > But we can't do that because webpack doesn't always set it? 

That one we can't use because it doesn't get updated when the minifier changes the code. It's `asset.info.contentHash` that isn't always set.

Comment (jsnajdr) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:204]: > That one we can't use because it doesn't get updated when the minifier changes the code.

I tested and debugged this a little and found something interesting. Chunk filenames are generated using the `chunk.contentHash` value, which is calculated from the unminified and unoptimized code. Not great. But then, one of the last asset processing stages is `PROCESS_ASSETS_STAGE_OPTIMIZE_HASH`, where the `RealContentHashPlugin` (enabled by default) does its work, and it updates the hashes, and renames the assets. These updated hashes are calculated from the minified code.

Our plugin runs in the `PROCESS_ASSETS_STAGE_ANALYSE`, which is after the hashes got updated. So in theory we should have access to them. But apparently they are calculated and updated only when the file name actually used the `[contenthash]` token? Or they are calculated always and we just haven't figured how to get them?

We don't use `[contenthash]` in file names, but we use it to generate a special `.php` asset which is then used by the WordPress script enqueueing engine to produce a `filename.js?ver=contenthash` script URL.

The one bad thing about this PR is that we're calculating the content hash ourselves, instead of letting webpack do it. I wish there was a way to fix that. Maybe there is really a webpack bug -- wanting to use the updated contenthash although it's not used in filenames is a reasonable demand.

Comment (anomiex) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:204]: > But apparently they are calculated and updated only when the file name actually used the `[contenthash]` token?

Yes, exactly. Which is why we can't take advantage of it being set here.

> The one bad thing about this PR is that we're calculating the content hash ourselves, instead of letting webpack do it.

Thus the `@todo` comment here. I hope we can avoid letting the perfect be the enemy of the good, and get this merged even though we have to calculate the hash outselves instead of trying to get Webpack to first develop a new feature.

Comment (jsnajdr) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:204]: If we have the goal to use the same hash as the real `[contenthash]`, and we are only forced to calculate it ourselves, they should match. I.e., if my webpack config uses a `[name].[contenthash].js` filename template, and also uses the extraction plugin, the `[contenthash]` and `version` should be the same. But they currently aren't. The differences I see are:
1) webpack uses `md4` algorithm by default, configurable with `output.hashFunction` option, while the plugin uses `sha512`.
2) webpack uses only the assets' contents to update the hash, while we hash the filename, too.
It's weird that even making these two changes, the hashes are still different for me.

It would be worthwhile to spend some timeboxed time trying to fix this before merging. Other than that, I think this patch is ready to land.

Comment (gziolo) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:204]: `md4` no longer works with Node 17+.

Replicating the same algorithm as webpack uses internally is rather complex. This is the initial commit when the support for `contentHash` was added:
https://github.com/webpack/webpack/commit/b929d4cdce433ac91fd8e4e5b8e6c8e90c6d5ca0.

Maybe we should try to detect `contentHash` first, and fall back to the custom handler otherwise. Example `contentHash` objects:

```js
{
  'css/mini-extract': '9dc6bf4629f53268df7c',
  javascript: 'bc28cb02479bf7449a77'
}
{ javascript: '684ed11ffa88cd017286' }
```

Comment (jsnajdr) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:204]: > `md4` no longer works with Node 17+.

webpack ships its own WASM module that implements `md4`. Instead of importing `createHash` from Node's `crypto` module, we can use the webpack version:
```js
const webpack = require( 'webpack' );
const hash = webpack.util.createHash( 'md4' );
```

> Replicating the same algorithm as webpack uses internally is rather complex.

The `RealContentHashPlugin` source is indeed very complex, and I'm not sure what set of possible use cases it covers, but I've been able to replicate the `contenthash` quite easily:
```js
const hash = createHash( 'md4' );
// asset.source.updateHash( hash );
hash.update( asset.source.buffer() );
const version = hash.digest( 'hex' );
```
Here the non-obvious step is to not use `asset.source.updateHash`, because that hashes a string `"RawSource" + source`, instead of just `source`. That's all the difference and when computing a content hash, we only want the source.

**Review by gziolo — APPROVED** [review]
As mentioned in my previous comment
https://github.com/WordPress/gutenberg/pull/34969#pullrequestreview-963009039, the proposed implementation works. Let's land it to remove the friction and explore improvements separately. @anomiex, thank you for providing the fix.

File: packages/dependency-extraction-webpack-plugin/lib/index.js (hunk 1 above)
Comment (anomiex) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:204]: > Here the non-obvious step is to not use `asset.source.updateHash`, because that hashes a string `"RawSource" + source`, instead of just `source`. That's all the difference and when computing a content hash, we only want the source.

Looking at it another way, we don't actually care what exactly goes into the hash as long as changes to the output result in changes to the hash, and non-changes to the output do not result in changes to the hash. Whether or not `"RawSource"` is incorporated in the hash makes no difference from that perspective.

OTOH, using `asset.source.updateHash` may have a performance advantage if the source is something like [ReplaceSource](https://github.com/webpack/webpack-sources/blob/main/lib/ReplaceSource.js) or [ConcatSource](https://github.com/webpack/webpack-sources/blob/main/lib/ConcatSource.js) that does non-trivial work in `source()` or `buffer()`.

Comment (jsnajdr) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:204]: > we don't actually care what exactly goes into the hash as long as changes to the output result in changes to the hash, and non-changes to the output do not result in changes to the hash.

I believe that my suggestions make an actual observable improvements to this behavior. I stopped hashing the file name, so renaming the file won't change the hash. In the end, the loading URL is going to be like `script.js?ver=hash`, where name change doesn't require a hash change to load the right asset.

Also, incorporating internal structure like `RawSource` tags changes the hash when webpack internals change, even if the content remains the same. It seems to me that `.updateHash` is more fit for internal purposes, not for calculating `contenthash`.

The performance advantage, I'm afraid it will never materialize. At some point webpack will write the asset to a file, and will call `source.buffer()` to construct the buffer to write. When calculating contenthash we'll just just construct the buffer a bit earlier, at a very late compilation stage where the source is unlikely to change further.
